tools/windows-node-installer/pkg/cloudprovider/azure/vm.go

```diff
@@ -0,0 +1,756 @@
+		return
+	}
+	networkProfile := vmStruct.VirtualMachineProperties.NetworkProfile
+	networkInterface := (*networkProfile.NetworkInterfaces)[0]
+	nicID := *networkInterface.ID
+	nicName := extractResourceName(nicID)
+	interfaceStruct, err := az.nicClient.Get(ctx, az.resourceGroupName, nicName, "")
+	if err != nil {
+		log.Error(err, fmt.Sprintf("cannot fetch the network interface data of %s", vmName))
+	}
+	interfacePropFormat := *(interfaceStruct.InterfacePropertiesFormat)
+	interfaceIPConfigs := *(interfacePropFormat.IPConfigurations)
+	ipConfigProp := *(interfaceIPConfigs[0].InterfaceIPConfigurationPropertiesFormat)
+	ipID := *(ipConfigProp.PublicIPAddress.ID)
+	ipName = extractResourceName(ipID)
+	return nil, ipName
+}
+
+// destroyIP deletes the IP address by taking it's name as argument.
+func (az *AzureProvider) destroyIP(ctx context.Context, ipName string) (err error) {
+	future, err := az.ipClient.Delete(ctx, az.resourceGroupName, ipName)
+	if err != nil {
+		log.Error(err, fmt.Sprintf("failed to delete the public IP: %s", ipName))
+		return
+	}
+	var response bool = false
+	var responseStatus int = 200
+	var r autorest.Response
+	for i := 1; i <= 300; i++ {
+		time.Sleep(1 * time.Second)
+		r, err = future.Result(az.ipClient)
+		response = r.IsHTTPStatus(responseStatus)
+		if response {
+			return nil
+		}
+	}
+	return err
+}
+
+// destroyInstance deletes the instance by taking it's name as argument.
+func (az *AzureProvider) destroyInstance(ctx context.Context, vmName string) (err error) {
+	future, err := az.vmClient.Delete(ctx, az.resourceGroupName, vmName)
+	if err != nil {
+		log.Error(err, fmt.Sprintf("failed to delete the instance: %s", vmName))
+		return
+	}
+	err = future.WaitForCompletionRef(ctx, az.vmClient.Client)
+	if err != nil {
+		log.Error(err, fmt.Sprintf("failed to delete the instance: %s", vmName))
+		return
+	}
+	return nil
+}
+
+// destroyNIC deletes the nic by taking it's name as argument.
+func (az *AzureProvider) destroyNIC(ctx context.Context, nicName string) (err error) {
+	future, err := az.nicClient.Delete(ctx, az.resourceGroupName, nicName)
+	if err != nil {
+		log.Error(err, fmt.Sprintf("failed to delete the nic: %s", nicName))
+		return
+	}
+	var response bool = false
+	var responseStatus int = 200
+	var r autorest.Response
+	for i := 1; i <= 300; i++ {
+		time.Sleep(1 * time.Second)
+		r, err = future.Result(az.nicClient)
+		response = r.IsHTTPStatus(responseStatus)
+		if response {
+			return nil
+		}
+	}
+	return err
+}
+
+// destroyNSG deletes the security groups by taking it's name as argument.
+func (az *AzureProvider) destroyNSG(ctx context.Context, nsgName string) (err error) {
+	future, err := az.nsgClient.Delete(ctx, az.resourceGroupName, nsgName)
+	if err != nil {
+		log.Error(err, fmt.Sprintf("failed to delete the sec group: %s", nsgName))
+		return
+	}
+	var response bool = false
+	var responseStatus int = 200
+	var r autorest.Response
+	for i := 1; i <= 300; i++ {
```

Comment:
The block is repeated in all the functions, do you want to write a function around this?

Reply:
In the azure sdk only for instance I could see fire and forget operation, but for other resources the only options I could think of atm is to continuously poll for them.

Reply:
Also I'm not sure to write for a helper function as the helper arguments for deleting the resources (ip, nic and secGroups) would be diferent.

Reply:
You don't get my point. Why do you need to wait for resources other than instance to be deleted to proceed? Will the presence of resource prevent another from being deleted?

Reply:
Yes a resource need to be deleted before proceeding to the next for ex:  I need to confirm the deletion of NIC before I go ahead and delete the IP resource.

Reply:
Are you saying that IP and NIC are connected? Even then you should be able to fire and forget deletion of IP. Same with storage.